### PR TITLE
JAMES-3028 Split blobstore cassandra

### PR DESCRIPTION
**Pull request by mbaechler**

### CODE REVIEW

**Comment by mbaechler** [comment]
Replace https://github.com/linagora/james-project/pull/3051

**Comment by Arsnael** [comment]
```
[6e9ee4d7ca090bd95773dc977428a78ea7442dd6] [ERROR] /james-project/server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStore.java:[114,55] unreachable statement
```

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/utils/DataChunker.java [hunk 1]
```diff
@@ -19,34 +19,57 @@
...
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
 
 public class DataChunker {
 
-    public Stream<Pair<Integer, ByteBuffer>> chunk(byte[] data, int chunkSize) {
+    public Flux<ByteBuffer> chunk(byte[] data, int chunkSize) {
         Preconditions.checkNotNull(data);
-        Preconditions.checkArgument(chunkSize > 0, "ChunkSize can not be negative");
+        Preconditions.checkArgument(chunkSize > 0, "ChunkSize can not be negative or 0");
```
Comment (nvduc91) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/utils/DataChunker.java]: IMO: **ChunkSize must be positive** is more fluent

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDumbBlobStore.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+    private Mono<Void> save(BucketName bucketName, BlobId blobId, Flux<ByteBuffer> chunksAsFlux) {
+        return saveBlobParts(bucketName, blobId, chunksAsFlux)
+            .flatMap(numberOfChunk -> saveBlobPartReference(bucketName, blobId, numberOfChunk));
+    }
+
+    private Mono<Integer> saveBlobParts(BucketName bucketName, BlobId blobId, Flux<ByteBuffer> chunksAsFlux) {
+        return chunksAsFlux
+            .index()
+            .concatMap(pair -> writePart(bucketName, blobId, pair.getT1().intValue(), pair.getT2()).thenReturn(getChunkNum(pair)))
```
Comment (rouazana) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDumbBlobStore.java]: cr/lf

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraDumbBlobStoreTest.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    static CassandraClusterExtension cassandraCluster = new CassandraClusterExtension(CassandraBlobModule.MODULE);
+
+    private DumbBlobStore testee;
+    private CassandraDefaultBucketDAO defaultBucketDAO;
+
+    @BeforeEach
+    void setUp(CassandraCluster cassandra) {
+        HashBlobId.Factory blobIdFactory = new HashBlobId.Factory();
+        CassandraBucketDAO bucketDAO = new CassandraBucketDAO(blobIdFactory, cassandra.getConf());
+        defaultBucketDAO = spy(new CassandraDefaultBucketDAO(cassandra.getConf()));
```
Comment (rouazana) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraDumbBlobStoreTest.java]: unneeded spy?

**Comment by mbaechler** [comment]
rebased on master

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDumbBlobStore.java (hunk 2 above)
Comment (mbaechler) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDumbBlobStore.java]: I did a bit more than that, I let you check that you are happy with the solution

**Comment by rouazana** [comment]
> rebased on master

bad autosquash :/

**Comment by mbaechler** [comment]
> > rebased on master
> 
> bad autosquash :/

sorry

**Comment by mbaechler** [comment]
Before:

Screenshot from 2020-02-10 09-26-50

Screenshot from 2020-02-09 12-35-20


After:

Screenshot from 2020-02-10 09-26-30

Screenshot from 2020-02-07 17-30-16

My analysis is: new implementation is a bit slower at enqueuing and a bit faster at dequeuing.

I would not bother to analyze that any further as the changes look really not significant.

WDYT?

**Comment by rouazana** [comment]
> I would not bother to analyze that any further as the changes look really not significant.

To analyze such differences we should at least launch each scenario 5 times. Agree to not investigate more for now.

**Comment by mbaechler** [comment]
```

[05326ae6d651d7a89386689cdab4fa3341000608] [ERROR] Errors: 
[05326ae6d651d7a89386689cdab4fa3341000608] [ERROR]   WebAdminUtilsTest.serverShouldBeAbletoStartConcurrently:38 » NotTerminated
```

test this please

**Comment by mbaechler** [comment]
```

[05326ae6d651d7a89386689cdab4fa3341000608] [ERROR] mixingSaveReadAndDeleteShouldReturnConsistentState  Time elapsed: 6.084 s  <<< ERROR!
[05326ae6d651d7a89386689cdab4fa3341000608] java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space
[05326ae6d651d7a89386689cdab4fa3341000608] Caused by: java.lang.OutOfMemoryError: Java heap space
[05326ae6d651d7a89386689cdab4fa3341000608] 
[05326ae6d651d7a89386689cdab4fa3341000608] [INFO] Running org.apache.james.blob.memory.MemoryBlobStoreTest
[05326ae6d651d7a89386689cdab4fa3341000608] [INFO] Tests run: 77, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.699 s - in org.apache.james.blob.memory.MemoryBlobStoreTest
```

test this please

**Comment by mbaechler** [comment]
same OOM

test this please

**Pull request merged**